accumulator/batchproof.go

```diff
@@ -112,122 +112,134 @@ func FromBytesBatchProof(b []byte) (BatchProof, error) {
+	// trees holds the entire proof tree of the batchproof in this way, sorted by the tuple[0].
+	trees := make([][3]node, 0, len(computablePositions))
+	// initialise the targetNodes for row 0.
+	// TODO: this would be more straight forward if bp.Proofs wouldn't contain the targets
+	proofHashes := make([]Hash, 0, len(proofPositions))
+	targets := bp.Targets
+	var targetsMatched uint64
+	for len(targets) > 0 {
+		if targets[0] == numLeaves-1 && numLeaves&1 == 1 {
+			// row 0 root.
+			rootCandidates = append(rootCandidates, node{Val: roots[0], Pos: targets[0]})
+			bp.Proof = bp.Proof[1:]
+			break
+		}
 
-	tagRow := bp.Targets
-	nextRow := []uint64{}
-	sortUint64s(tagRow) // probably don't need to sort
+		if uint64(len(proofPositions)) > targetsMatched &&
+			targets[0]^1 == proofPositions[targetsMatched] {
+			lr := targets[0] & 1
+			targetNodes = append(targetNodes, node{Pos: targets[0], Val: bp.Proof[lr]})
+			proofHashes = append(proofHashes, bp.Proof[lr^1])
+			targetsMatched++
+			bp.Proof = bp.Proof[2:]
+			targets = targets[1:]
+			continue
+		}
 
-	// TODO it's ugly that I keep treating the 0-row as a special case,
-	// and has led to a number of bugs.  It *is* special in a way, in that
-	// the bottom row is the only thing you actually prove and add/delete,
-	// but it'd be nice if it could all be treated uniformly.
+		if len(bp.Proof) < 2 || len(targets) < 2 {
+			return false, nil, nil
+		}
 
-	if verbose {
-		fmt.Printf("tagrow len %d\n", len(tagRow))
+		targetNodes = append(targetNodes,
+			node{Pos: targets[0], Val: bp.Proof[0]},
+			node{Pos: targets[1], Val: bp.Proof[1]})
+		bp.Proof = bp.Proof[2:]
+		targets = targets[2:]
 	}
 
-	var left, right uint64
-
-	// iterate through rows
-	for row := uint8(0); row <= rows; row++ {
-		// iterate through tagged positions in this row
-		for len(tagRow) > 0 {
-			// Efficiency gains here. If there are two or more things to verify,
-			// check if the next thing to verify is the sibling of the current leaf
-			// we're on. Siblingness can be checked with bitwise XOR but since targets are
-			// sorted, we can do bitwise OR instead.
-			if len(tagRow) > 1 && tagRow[0]|1 == tagRow[1] {
-				left = tagRow[0]
-				right = tagRow[1]
-				tagRow = tagRow[2:]
-			} else { // if not only use one tagged position
-				right = tagRow[0] | 1
-				left = right ^ 1
-				tagRow = tagRow[1:]
-			}
+	proofHashes = append(proofHashes, bp.Proof...)
+	bp.Proof = proofHashes
+
+	// hash every target node with its sibling (which either is contained in the proof or also a target)
+	for len(targetNodes) > 0 {
+		var target, proof node
+		target = targetNodes[0]
+		if len(proofPositions) > 0 && target.Pos^1 == proofPositions[0] {
+			// target has a sibling in the proof positions, fetch proof
+			proof = node{Pos: proofPositions[0], Val: bp.Proof[0]}
+			proofPositions = proofPositions[1:]
+			bp.Proof = bp.Proof[1:]
+			targetNodes = targetNodes[1:]
+			goto hash
```

Comment:
Could be abstracted into a separate function instead of a goto?

Reply:
I think the goto is simpler here because a separate function would have to take a ton of arguments and return all these slices i am appending to. It's also a bit easier to read imo than a function call in this case because it just jumps 10 lines down.
I think this is one of those rare cases where a goto is ok but if you strongly disagree i can change it.

Reply:
How about just an `else`?  That seems even easier to understand

```
			proofPositions = proofPositions[1:]
			bp.Proof = bp.Proof[1:]
			targetNodes = targetNodes[1:]
		} else {
         		// target should have its sibling in targetNodes
	        	if len(targetNodes) == 1 {
		        	// sibling not found
	        		return false, nil, nil
	        	}
        		proof = targetNodes[1]
	        	targetNodes = targetNodes[2:]
		}
		left := target
		right := proof
```

Reply:
changed it to and `else`.